source/libraries/CompileSolidity.ts

```diff
@@ -11,58 +11,66 @@ interface CompileContractsOutput {
     output?: string;
 }
 
-class SolidityContractCompiler {
+export class SolidityContractCompiler {
     private contractInputDirectoryPath: string;
     private contractOutputDirectoryPath: string;
     private contractOutputFileName: string;
 
     public constructor(contractInputDirectoryPath: string, contractOutputDirectoryPath: string, contractOutputFileName: string) {
+        if (contractInputDirectoryPath.lastIndexOf(path.sep) != contractInputDirectoryPath.length - 1) {
+            contractInputDirectoryPath += path.sep;
+        }
+        if (contractOutputDirectoryPath.lastIndexOf(path.sep) != contractOutputDirectoryPath.length - 1) {
+            contractOutputDirectoryPath += path.sep;
+        }
         this.contractInputDirectoryPath = contractInputDirectoryPath;
         this.contractOutputDirectoryPath = contractOutputDirectoryPath;
         this.contractOutputFileName = contractOutputFileName;
     }
 
-    public async compileContracts(): Promise<CompileContractsOutput> {
+    public readCallback(path: string): { contents?: string, error?: string } {
         try {
-            // Compile all contracts in the specified input directory
-            const compilerInputJson = await this.generateCompilerInput();
-            const compilerOutputJson: string = compileStandardWrapper(JSON.stringify(compilerInputJson), this.readCallback);
-            const compilerOutput: CompilerOutput = JSON.parse(compilerOutputJson);
-            if (compilerOutput.errors) {
-                let errors = "";
-                for (let error of compilerOutput.errors) {
-                    errors += error.formattedMessage + "\n";
-                }
-                throw new Error("The following errors/warnings were returned by solc:\n\n" + errors);
-            }
+            const result = fs.readFileSync(path, 'utf8');
+            return { contents: result };
+        } catch (error) {
+            return { error: error.message };
+        }
+    }
 
-            // Create output directory (if it doesn't exist)
-            mkdirp(this.contractOutputDirectoryPath, this.mkdirpCallback);
+    public async compileContracts(): Promise<CompileContractsOutput> {
+        // Check if all contracts are cached (and thus do not need to be compiled)
+        const stats = fs.statSync(this.contractOutputDirectoryPath + this.contractOutputFileName);
+        const lastCompiledTimestamp = stats.mtime;
+        const uncachedFiles = await recursiveReadDir(this.contractInputDirectoryPath, [function(file: string, stats: fs.Stats): boolean {return stats.isDirectory() || (stats.isFile() && path.extname(file) != ".sol") || (stats.isFile() && path.extname(file) == ".sol" && stats.mtime < lastCompiledTimestamp);}]);
```

Comment:
Consider extracting the ignore function out into a separate function to improve readability.